### PR TITLE
Add Diagonal Assembly for AtPoints

### PR DESCRIPTION
**Pull request by jeremylt**
Adding diagonal assembly for AtPoints operators by probing each element with unit vectors

- [x] CPU impl
- [x] Tests

### CODE REVIEW

File: examples/petsc/bpsswarm.c [hunk 1]
```diff
@@ -302,14 +303,9 @@ int main(int argc, char **argv) {
...
       PetscCall(PCSetType(pc, PCJACOBI));
-      PetscCall(PCJacobiSetType(pc, PC_JACOBI_ROWSUM));
+      PetscCall(PCJacobiSetType(pc, PC_JACOBI_DIAGONAL));
     } else {
       PetscCall(PCSetType(pc, PCNONE));
-      MatNullSpace nullspace;
-
-      PetscCall(MatNullSpaceCreate(PETSC_COMM_WORLD, PETSC_TRUE, 0, 0, &nullspace));
-      PetscCall(MatSetNullSpace(mat_O, nullspace));
-      PetscCall(MatNullSpaceDestroy(&nullspace));
```
Comment (jeremylt) [inline examples/petsc/bpsswarm.c:312]: I made the two match. Alternatively, we could add this to `bps.c`?

Comment (zatkins-dev) [reply examples/petsc/bpsswarm.c:312]: Ah, I think I copied from `bpssphere.c` originally -- if this isn't needed here I'm good with removing.

File: backends/ref/ceed-ref-operator.c [hunk 2]
```diff
@@ -1109,8 +1109,212 @@ static int CeedOperatorLinearAssembleQFunctionAtPointsUpdate_Ref(CeedOperator op
...
+    CeedCallBackend(CeedElemRestrictionApplyAtPointsInElement(rstr_points, e, CEED_NOTRANSPOSE, point_coords, impl->point_coords_elem, request));
+    CeedCallBackend(CeedElemRestrictionGetNumPointsInElement(rstr_points, e, &num_points));
+
+    // Input basis apply for non-active bases
+    CeedCallBackend(CeedOperatorInputBasisAtPoints_Ref(e, num_points_offset, num_points, qf_input_fields, op_input_fields, num_input_fields, in_vec,
+                                                       impl->point_coords_elem, true, e_data, impl, request));
+
+    // Loop over points on element
+    e_vec_size = (is_active_at_points ? num_points : elem_size_active) * num_comp_active;
+    for (CeedInt s = 0; s < e_vec_size; s++) {
```
Comment (zatkins-dev) [inline backends/ref/ceed-ref-operator.c:1192]: Just to clarify, for our typical Jacobian operator at points, this is a loop over the nodes of each element that computes the diagonal contributions for that node from the points within that element, right?

I want to make sure we aren't doing something with like `O(num_points^2)`  complexity.

**Review by zatkins-dev — APPROVED** [review]
Looks good to me, thanks! Have you looked at any performance comparison between the old implementation and this one?

File: backends/ref/ceed-ref-operator.c (hunk 2 above)
Comment (jeremylt) [reply backends/ref/ceed-ref-operator.c:1192]: For our typical approach, we assemble the QF by repeated application of D [1, 0, 0...] and then compute the matrix triple product B D B^T and extract the element operator matrix diagonal. (we might be doing something clever there so we don't compute the full matrix triple product)

Here we are instead repeatedly applying B D B^T [1, 0, 0...] to build the element operator matrix diagonal.

This PR should be more expensive than our typical approach when we have consistent quadrature points in each element but less expensive than building B for every element

Comment (zatkins-dev) [reply backends/ref/ceed-ref-operator.c:1192]: Gotcha, sounds good to me.

**Pull request merged**